src/Storages/FileLog/StorageFileLog.cpp

```diff
@@ -0,0 +1,432 @@
+#include <Interpreters/Context.h>
+#include <Interpreters/InterpreterInsertQuery.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTExpressionList.h>
+#include <Parsers/ASTInsertQuery.h>
+#include <Parsers/ASTLiteral.h>
+#include <Processors/Executors/PullingPipelineExecutor.h>
+#include <Processors/Pipe.h>
+#include <Processors/Sources/SourceFromInputStream.h>
+#include <Storages/FileLog/FileLogSource.h>
+#include <Storages/FileLog/ReadBufferFromFileLog.h>
+#include <Storages/FileLog/StorageFileLog.h>
+#include <Storages/StorageFactory.h>
+#include <Storages/StorageMaterializedView.h>
+#include <Common/Exception.h>
+#include <Common/Macros.h>
+#include <Common/formatReadable.h>
+#include <Common/getNumberOfPhysicalCPUCores.h>
+#include <Common/quoteString.h>
+#include <Common/setThreadName.h>
+#include <Common/typeid_cast.h>
+#include <common/logger_useful.h>
+#include <common/sleep.h>
+
+#include <Poco/DirectoryIterator.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int BAD_ARGUMENTS;
+}
+
+namespace
+{
+    const auto RESCHEDULE_MS = 500;
+    const auto MAX_THREAD_WORK_DURATION_MS = 60000;  // once per minute leave do reschedule (we can't lock threads in pool forever)
+}
+
+StorageFileLog::StorageFileLog(
+    const StorageID & table_id_,
+    ContextPtr context_,
+    const ColumnsDescription & columns_,
+    const String & relative_path_,
+    const String & format_name_,
+    std::unique_ptr<FileLogSettings> settings)
+    : IStorage(table_id_)
+    , WithContext(context_->getGlobalContext())
+    , filelog_settings(std::move(settings))
+    , path(getContext()->getUserFilesPath() + "/" + relative_path_)
+    , format_name(format_name_)
+    , log(&Poco::Logger::get("StorageFileLog (" + table_id_.table_name + ")"))
+{
+    StorageInMemoryMetadata storage_metadata;
+    storage_metadata.setColumns(columns_);
+    setInMemoryMetadata(storage_metadata);
+
+    if (std::filesystem::is_regular_file(path))
+    {
+        auto normal_path = std::filesystem::path(path).lexically_normal().native();
```

Comment:
btw why `lixically_normal` and not `canonical` ? but it does not matter here probably